demo-circuit/src/blaze_csw/constraints/data_structures.rs

```diff
@@ -0,0 +1,1496 @@
+    where
+        F: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<CswUtxoInputData>,
+    {
+        let (output, secret_key) = match f() {
+            Ok(csw_utxo_input_data) => {
+                let csw_utxo_input_data = csw_utxo_input_data.borrow().clone();
+                (
+                    Ok(csw_utxo_input_data.output),
+                    Ok(csw_utxo_input_data.secret_key),
+                )
+            }
+            _ => (
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+            ),
+        };
+
+        let output_g = CswUtxoOutputDataGadget::alloc(cs.ns(|| "alloc output"), || output)?;
+
+        let secret_key_g = Vec::<Boolean>::alloc(cs.ns(|| "alloc secret key"), || secret_key)?
+            .try_into()
+            .map_err(|_| {
+                SynthesisError::Other(format!(
+                    "invalid size for secret_key, expected {} bits",
+                    SIMULATED_SCALAR_FIELD_MODULUS_BITS
+                ))
+            })?;
+
+        Ok(Self {
+            output_g,
+            secret_key_g,
+        })
+    }
+
+    fn alloc_input<F, T, CS: ConstraintSystemAbstract<FieldElement>>(
+        _cs: CS,
+        _f: F,
+    ) -> Result<Self, SynthesisError>
+    where
+        F: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<CswUtxoInputData>,
+    {
+        unimplemented!()
+    }
+}
+
+pub struct CswUtxoProverDataGadget {
+    pub input_g: CswUtxoInputDataGadget, // unspent output we are trying to withdraw
+    pub mst_path_to_output_g: GingerMHTBinaryGadget, // path to output in the MST of the known state
+}
+
+impl CswUtxoProverDataGadget {
+    /// Enforce that:
+    /// 1) H(self.input_g.output_g) belongs to merkle tree with root 'scb_new_mst_root_g';
+    /// 2) H(self.input_g.output_g) == 'nullifier_g'
+    /// 3) self.input_g.output_g.amount_g == 'amount_g'
+    pub(crate) fn conditionally_enforce_utxo_withdrawal<
+        CS: ConstraintSystemAbstract<FieldElement>,
+    >(
+        &self,
+        mut cs: CS,
+        scb_new_mst_root_g: &FieldElementGadget,
+        nullifier_g: &FieldElementGadget,
+        amount_g: &FieldElementGadget,
+        should_enforce: &Boolean,
+    ) -> Result<(), SynthesisError> {
+        // Enfore UTXO output hash computation
+        let box_type_coin_g = FieldElementGadget::from(
+            cs.ns(|| "alloc BoxType.Coin constant"),
+            &FieldElement::from(BoxType::CoinBox as u8),
+        );
+
+        let mut output_hash_elements_g = self
+            .input_g
+            .output_g
+            .to_field_gadget_elements(cs.ns(|| "alloc output hash elements"))?;
+
+        debug_assert_eq!(output_hash_elements_g.len(), 3);
+        output_hash_elements_g.push(box_type_coin_g);
+
+        let output_hash_g = FieldHashGadget::enforce_hash_constant_length(
+            cs.ns(|| "H(input.output)"),
+            &output_hash_elements_g,
+        )?;
```

Comment:
As described in [snarks-architecture/issues/3](https://github.com/HorizenLabs/snarks-architecture/issues/3), the current approach to separate utxo/ft nullifiers is a little bit "dangerous", as its security depends on the type of data to be hashed. Let us use Poseidon's personalization feature to circumvent this. A choice of different personalization string/seeds guarantees that the resulting hash functions used for computing nullifiers are ensured to be uncorrelated. (The same is true if personalization is used for only one of the hashes). As domain separation is a useful feature for a transparent design of cryptographic protocols in general, I propose to

- implement a personalization gadget for Poseidon, which directly sets the internal state to the result of the personalization function from the primitive. This gadget does not cost any constraints.

Then, apply 
-  Poseidon with personalization string such as `seed = "csw_utxo"` for computing the nullifier of utxos
`nullifier = PH_seed(input.output)`, and 
- Poseidon without personalization for the nullifier of forward transactions, `nullifier = PH(ft_input)`.